scaleway/resource_iot_hub_beta.go

```diff
@@ -0,0 +1,249 @@
+		OrganizationID: d.Get("organization_id").(string),
+		ProductPlan:    iot.ProductPlan(d.Get("product_plan").(string)),
+	}
+
+	res, err := iotAPI.CreateHub(req)
+	if err != nil {
+		return err
+	}
+
+	err = waitIotHub(iotAPI, region, res.ID, iot.HubStatusReady)
+	if err != nil {
+		return err
+	}
+
+	// Enable/Disable hub if needed. The only case we need to check is to eventually
+	// disable the hub (enabled by default)
+	// We need to ensure the hub was fully enabled before because the hub cannot
+	// be updated while enabling.
+	if enabled, ok := d.GetOk("enabled"); ok && !enabled.(bool) {
+		_, err = iotAPI.DisableHub(&iot.DisableHubRequest{
+			Region: region,
+			HubID:  res.ID,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = waitIotHub(iotAPI, region, res.ID, iot.HubStatusDisabled)
+		if err != nil {
+			return err
+		}
+	}
+
+	d.SetId(newRegionalId(region, res.ID))
+
+	return resourceScalewayIotHubBetaRead(d, m)
+}
+
+func resourceScalewayIotHubBetaRead(d *schema.ResourceData, m interface{}) error {
+	iotAPI, region, hubID, err := iotAPIWithRegionAndID(m, d.Id())
+	if err != nil {
+		return err
+	}
+
+	////
+	// Read Hub
+	////
+	response, err := iotAPI.GetHub(&iot.GetHubRequest{
+		Region: region,
+		HubID:  hubID,
+	})
+	if err != nil {
+		if is404Error(err) {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	_ = d.Set("region", string(region))
+	_ = d.Set("name", response.Name)
+	_ = d.Set("status", response.Status.String())
+	_ = d.Set("product_plan", response.ProductPlan.String())
+	_ = d.Set("endpoint", response.Endpoint)
+	_ = d.Set("created_at", response.CreatedAt)
+	_ = d.Set("updated_at", response.UpdatedAt)
+	_ = d.Set("enabled", response.Enabled)
+	_ = d.Set("device_count", response.DeviceCount)
+	_ = d.Set("connected_device_count", response.ConnectedDeviceCount)
+
+	return nil
+}
+
+func resourceScalewayIotHubBetaUpdate(d *schema.ResourceData, m interface{}) error {
+	iotAPI, region, hubID, err := iotAPIWithRegionAndID(m, d.Id())
+	if err != nil {
+		return err
+	}
+
+	////
+	// Construct UpdateHubRequest
+	////
+	updateRequest := &iot.UpdateHubRequest{
+		Region: region,
+		HubID:  hubID,
+	}
+
+	if d.HasChange("name") {
+		updateRequest.Name = scw.StringPtr(d.Get("name").(string))
+	}
+
+	////
+	// Apply Update
+	////
+	_, err = iotAPI.UpdateHub(updateRequest)
+	if err != nil {
+		return err
+	}
+
+	////
+	// Enable/Disable hub if needed
+	////
+	if d.HasChange("enabled") {
+		newTargetStatus := d.Get("enabled").(bool)
+
+		var err error
+		if newTargetStatus == true {
+			_, err = iotAPI.EnableHub(&iot.EnableHubRequest{
+				Region: region,
+				HubID:  hubID,
+			})
+			if err != nil {
+				return err
+			}
+
+			err = waitIotHub(iotAPI, region, hubID, iot.HubStatusReady)
+			if err != nil {
+				return err
+			}
+		} else {
+			_, err = iotAPI.DisableHub(&iot.DisableHubRequest{
+				Region: region,
+				HubID:  hubID,
+			})
+			if err != nil {
```

Comment:
You don't wait here ?

Reply:
There can be an specific path (even if it would be hard to produce in terraform) where this could lead to a "Resource is busy", indeed. I changed this to wait in every case. This was not done before because the devices are disconnected right during the call :)